tsh: ignore empty or non-existing config files

marcoandredinis opened the pull request:
If the newly created config.yaml didnt exist we would load the default
values and continue the flow
However, we were not resetting the `err`'s value and would have an ERROR
message at the end and an invalid exit code

Most of the commands would reset that variable to the output of the
command's execution
One of them was not: `tsh version`
The version command has no return value, so the program would execute
as expected until the last statement: `trace.Wrap(error)` which would
re-use the `err` variable whose value is the result of the `loadConfig`
method.

We could either reset the `err`s value inside the `PrintVersion` switch
case block or reset it right after we check for `IsNotFound`.

We ended up picking the first option as it seems cleaner

```
 # before
$ make full > /dev/null; build/tsh version
Teleport v10.0.0-dev git:v8.0.0-alpha.1-899-g335adf1f4 go1.18
ERROR: open /home/marco/.tsh/config/config.yaml: no such file or directory

 # after
$ make full > /dev/null; build/tsh version
Teleport v10.0.0-dev git:v8.0.0-alpha.1-899-g335adf1f4 go1.18
```

r0mant commented on tool/tsh/tsh.go:
```diff
@@ -689,7 +689,10 @@ func Run(args []string, opts ...cliOption) error {
 	setEnvFlags(&cf, os.Getenv)
 
 	confOptions, err := loadConfig(cf.HomePath)
-	if err != nil && !trace.IsNotFound(err) {
+	if trace.IsNotFound(err) {
+		err = nil
+	}
```
Hm, I wonder if it would be a little cleaner interface if we made `loadConfig` return an empty config in case the config file does not exist. Then the caller won't have to worry about handling the not found error. cc @lxea

lxea replied:
That makes sense to me

r0mant reviewed (commented):
@marcoandredinis Good catch! Have one comment.

r0mant reviewed (commented):
@marcoandredinis Also, please include test coverage with the change. If you take my suggestion, I think you can just write a test for `loadConfig` and make sure it doesn't return error if the file isn't found.

marcoandredinis committed:
tsh: don't propagate error on config notfound

If the newly created config.yaml doesn't exist we load the default
values and continue the flow
However, we were not resetting the `err`'s value and would have an ERROR
message at the end and an invalid exit code

Most of the commands would reset that variable to the output of the
command's execution
One of them was not: `tsh version`
The version command has not return value, so the program would execute
as expected until the last statement: `trace.Wrap(error)` which would
re-use the `err` variable whose value is the result of the `loadConfig`
method.

We could clean up the `err`'s value or return no error at all and an
empty config when the file doesn't exist

We ended up picking the latter option as it seems cleaner

```
 # before
$ make full > /dev/null; build/tsh version
Teleport v10.0.0-dev git:v8.0.0-alpha.1-899-g335adf1f4 go1.18
ERROR: open /home/marco/.tsh/config/config.yaml: no such file or directory

 # after
$ make full > /dev/null; build/tsh version
Teleport v10.0.0-dev git:v8.0.0-alpha.1-899-g335adf1f4 go1.18
```

marcoandredinis commented:
Should we also get a valid config if file is empty? :thinking:

r0mant reviewed (approved):
> Should we also get a valid config if file is empty?

@marcoandredinis If yaml decoder fails with empty file, I'd return an empty config in that case also, yes. Otherwise lgtm.

marcoandredinis committed:
return empty config when file is empty

marcoandredinis committed:
dont assume EOF is always empty file

r0mant commented on tool/tsh/tshconfig.go:
```diff
@@ -46,14 +47,28 @@ type ExtraProxyHeaders struct {
+	if err != nil {
+		return nil, trace.ConvertSystemError(err)
+	}
+	if len(bs) == 0 {
```
Nit: I'd use bytes.TrimSpace before comparing the length just in case.

marcoandredinis replied:
I ended up removing the empty/only spaces checks
yaml.Unmarshal handles that just fine

We initially added the check because we were decoding using `yaml.NewDecoder(configFile).Decode(&cfg)` and that requires at least one valid node: https://pkg.go.dev/gopkg.in/yaml.v2#Decoder.Decode
However, the Unmarshal is fine with empty input

r0mant commented on tool/tsh/tshconfig.go:
```diff
@@ -46,14 +47,28 @@ type ExtraProxyHeaders struct {
 	cfg := TshConfig{}
-	err = yaml.NewDecoder(configFile).Decode(&cfg)
-	return &cfg, trace.Wrap(err)
+	if yaml.Unmarshal(bs, &cfg); err != nil {
```
```suggestion
	if err := yaml.Unmarshal(bs, &cfg); err != nil {
```

r0mant commented on tool/tsh/tshconfig.go:
```diff
@@ -46,14 +47,28 @@ type ExtraProxyHeaders struct {
 	}
 	defer configFile.Close()
+
+	bs, err := io.ReadAll(configFile)
```
Why not just use `os.ReadFile` instead of Open+ReadAll then?

r0mant reviewed (changes requested):
Have some comments after the latest changes.

marcoandredinis committed:
fix err handling in yaml unmarshal

yaml.Unmarshal handles empty files just fine, no need to do prior checks for
empty or space only files

lxea reviewed (approved):
Thanks!

r0mant reviewed (approved):
@marcoandredinis Let's merge and backport to branch/v9 please.

marcoandredinis commented:
@zmb3 Indeed
On it
